Matplotlib sharon

sharonyang opened the pull request:
Have not yet put together a class that display multiple devices' stats at the same time. But this class Grapher can plot individual Device one at a time.

sharonyang committed:
Added a PlotTool class to help graphing.

This will probably be combined with Statistics class to enable
the graphing feature of the class. This is not done yet.

sharonyang committed:
Modified Grapher class.

Usage: instantiate a Grapher class with a Statistics object, then
we can graph any Device (link, flow, host). What we need at this
moment is to figure out the best way to lay out multiple Device's
on the same screen.

lakshbhasin commented:
I haven't looked too closely, but I wanted to remind you that the plots should look like those in http://courses.cms.caltech.edu/cs143/Project/NetworkSimTestCases-2015.pdf . So with multiple links/flows plotted on the same axes. For example, you will probably want to plot all flows' rates on the same subplot. See page 7 of the PDF (not sure if we need to plot window size yet though...)

sharonyang committed:
Made changes to plotting output (with legends) based on class notes.

The class note is here:
http://courses.cms.caltech.edu/cs143/Project/NetworkSimTestCases
-2015.pdf
Now the functions take in a dictionary as an input to plot. We need
to fix it once this PR is LGTM'ed.

sharonyang committed:
Combined three graphing functions into one for graphing the entire
network.

No longer separates each device (link, flow, host) in individual
windows. Have one overall panels to contain all devices.

sharonyang committed:
Modified graphic layout per Laksh's comment in team chat.

This includes changing the display options and adding a window size
over global time graph for flows.

sharonyang committed:
Updated Grapher class based on Laksh's comments regarding to window
sizes.

sharonyang committed:
Finalized graphing class for PR.

This includes getting rid of dictionary as input and including the
graphing feature in the main event loop.

sharonyang committed:
Modified Packet Losses Graph to compute interval of aggregated loss.

Also modified plot titles per Laksh's comments.

sharonyang committed:
Merge branch 'master' of https://github.com/lakshbhasin/network_simulator into matplotlib-sharon

lakshbhasin commented on network_simulator/plot_tool.py:
```diff
@@ -0,0 +1,300 @@
+import matplotlib.pyplot as plt
+import numpy as np
+
+from statistics import *
```
Please don't use the "from X import *" format (outside of "from common import *") in any project files anymore, as it can create nasty circular dependencies. Just import the things you need (check for errors in PyCharm). I have already fixed this in the team-debug branch, so you can just change this in the files you've written.

sharonyang replied:
Done.

lakshbhasin commented on network_simulator/plot_tool.py:
```diff
@@ -0,0 +1,300 @@
+
+    def gen_count_interval_list(self, lst):
+        """
+        Computes a tuple list of counts based on a timestamp list.
```
You should mention that windowing is used to group together data that occurred between a given timestamp and the previous timestamp in the output list (where the timestamps are regularly spaced apart in the output list).

sharonyang replied:
Done.

lakshbhasin commented on network_simulator/plot_tool.py:
```diff
@@ -0,0 +1,300 @@
+        Computes a tuple list of counts based on a timestamp list.
+
+        :param list lst: list of timestamps to be converted.
+        :return: list
```
State that this is a list of (time, count) tuples and describe what the counts would represent.

sharonyang replied:
Done.

lakshbhasin commented on network_simulator/plot_tool.py:
```diff
@@ -0,0 +1,300 @@
+        :return: list
+        """
+        curr_count = 0
+        curr_time = GRAPH_WINDOW_SIZE
```
Refactor (shift+F6 in PyCharm on Linux) to "curr_window_end_time", since it seems like this is the endpoint of the windows you are using. Could also describe this in a comment, in addition to renaming.

sharonyang replied:
Done.

lakshbhasin commented on network_simulator/plot_tool.py:
```diff
@@ -0,0 +1,300 @@
+        :param list lst: list of timestamps to be converted.
+        :return: list
+        """
+        curr_count = 0
```
Refactor to curr_window_count.

sharonyang replied:
Done.

lakshbhasin commented on network_simulator/plot_tool.py:
```diff
@@ -0,0 +1,300 @@
+                output.append((curr_time, curr_count))
+                curr_count = 0
+                curr_time += GRAPH_WINDOW_SIZE
+            if timestamp <= curr_time:
```
nit: this could just be an "else:"

sharonyang replied:
It should be continuous after one reset and therefore should be an "if"?

sharonyang replied:
It should be continuous after one reset and therefore should be an "if"?

lakshbhasin commented on network_simulator/main_event_loop.py:
```diff
@@ -84,4 +85,7 @@ def run(self):
 
         logger.info("Finished running main Event loop.")
 
-        # TODO(team): Output final Statistics
```
There is another TODO to output final Statistics a few lines above (when a warning occurs).

sharonyang replied:
Done.

lakshbhasin commented on network_simulator/plot_tool.py:
```diff
@@ -0,0 +1,300 @@
+        :return: list
+        """
+        curr_count = 0
+        curr_sum = 0
```
curr_count, curr_sum, and curr_time are not descriptive enough names.

lakshbhasin commented on network_simulator/plot_tool.py:
```diff
@@ -0,0 +1,300 @@
+        list.
+
+        :param list tuple_list: list of tuples to be converted.
+        :return: list
```
What is the format of the list returned?

lakshbhasin commented on network_simulator/plot_tool.py:
```diff
@@ -0,0 +1,300 @@
+
+    def gen_rate_tuple_list(self, tuple_list):
+        """
+        Computes a tuple list of rates based on a (timestamp, data) tuple
```
Mention that the data always represents packet sizes (right?). Could just replace "data" with packet_size and refer to these rates as having units of bits per second.

sharonyang replied:
Done.

lakshbhasin commented on network_simulator/plot_tool.py:
```diff
@@ -0,0 +1,300 @@
+            timestamp, data = tuple_list[tup_idx]
+            if timestamp > curr_time:
+                if curr_count != 0:
+                    output.append((curr_time, float(curr_sum) / curr_count))
```
This doesn't make sense to me. You should be taking the sum of the packet sizes in an interval (in bits) and then dividing by the window size of the interval (in sec), which is GRAPH_WINDOW_SIZE. This will give an output in units of bits per second. I don't see why you need curr_count at all.

sharonyang replied:
Done.

lakshbhasin commented on network_simulator/plot_tool.py:
```diff
@@ -0,0 +1,300 @@
+
+from statistics import *
+
+class PlotTool(object):
```
It seems like most of the functions in this class are not instance functions, i.e. they are static functions. You can label these with @staticmethod before the function signature, and can get rid of the self in the arguments. Also, get rid of the __init__ function. Change all usages to just call PlotTool.func().

sharonyang replied:
Done.

lakshbhasin commented on network_simulator/plot_tool.py:
```diff
@@ -0,0 +1,300 @@
+            return
+
+        # Plot 1-D different values on a x-y plane.
+        plt.plot(lst, np.zeros_like(lst), marker='o', linestyle='--')
```
I have no idea why this is needed. All plots use both the x and y axes.

sharonyang replied:
Done.

lakshbhasin commented on network_simulator/plot_tool.py:
```diff
@@ -0,0 +1,300 @@
+        Graphs the link stats of each available link.
+        """
+        # Set up legend storage for link names.
+        occpy_legend = []
```
Have a comment explaining what each legend is for, before each declaration.

sharonyang replied:
Done.

lakshbhasin commented on network_simulator/plot_tool.py:
```diff
@@ -0,0 +1,300 @@
+
+            # Packet loss times.
+            plt.subplot(434)
+            loss_rate_lst = self.tool.gen_count_interval_list(
```
this is actually a "loss_count_list".

sharonyang replied:
Done.

lakshbhasin commented on network_simulator/plot_tool.py:
```diff
@@ -0,0 +1,300 @@
+        """
+        Graphs the content of a tuple list.
+
+        :param list tuple_list: if we have a newer version of the list,
```
Not sure what this comment is referring to. Isn't this just the list being plotted?

sharonyang replied:
Done.

lakshbhasin commented on network_simulator/plot_tool.py:
```diff
@@ -0,0 +1,300 @@
+
+    def graph_tuple_list(self, tuple_list=[], scatter=False):
+        """
+        Graphs the content of a tuple list.
```
How many dimensions are in the tuple? 2D?

sharonyang replied:
Done.

lakshbhasin commented on network_simulator/plot_tool.py:
```diff
@@ -0,0 +1,300 @@
+            self.tool.graph_tuple_list(loss_rate_lst)
+            loss_legend.append(link)
+
+            # Packet transmission w.r.t time.
```
Don't graph packet transmission vs time. What we want for the third graph is the link flow rate vs time. This is a rate that must be calculated through windowing.

sharonyang replied:
Done.

lakshbhasin commented on network_simulator/plot_tool.py:
```diff
@@ -0,0 +1,300 @@
+        plt.legend(trans_legend)
+        # Set up labels.
+        plt.xlabel("Time (sec)")
+        plt.ylabel("Size of Packet Transmitted (Bits)")
```
Again, third graph for links should be the link flow rate in Mbps vs time. To put it in Mbps, you will have to take the rate tuple list and divide its rates by 1024^2, i.e. the MEGABIT constant in common.py (Yubo will fix the value of this constant so it is in base 2 soon).

sharonyang replied:
Done.

lakshbhasin commented on network_simulator/plot_tool.py:
```diff
@@ -0,0 +1,300 @@
+        plt.xlabel("Time (sec)")
+        plt.ylabel("Buffer Occupancy (pkts)")
+        # Set up grid on graph.
+        plt.grid(True)
```
Don't use plt.grid anywhere. Grids are distracting and often serve no useful purpose in scientific graphs. matplotlib will automatically include tick marks for major intervals on its axes.

sharonyang replied:
Done.

lakshbhasin commented on network_simulator/plot_tool.py:
```diff
@@ -0,0 +1,300 @@
+
+    def graph_links(self):
+        """
+        Graphs the link stats of each available link.
```
State what these stats include.

sharonyang replied:
Done.

lakshbhasin commented on network_simulator/plot_tool.py:
```diff
@@ -0,0 +1,300 @@
+
+    def graph_hosts(self):
+        """
+        Graphs the host stats of each available host.
```
State what these stats are.

sharonyang replied:
Done.

lakshbhasin commented on network_simulator/plot_tool.py:
```diff
@@ -0,0 +1,300 @@
+
+    def graph_flows(self):
+        """
+        Graphs the flow stats of each available flow.
```
State what these stats are.

sharonyang replied:
Done.

lakshbhasin commented on network_simulator/plot_tool.py:
```diff
@@ -0,0 +1,300 @@
+        Graphs the flow stats of each available flow.
+        """
+        # Set up legend storage for flow names.
+        sent_legend = []
```
Describe legends before each declaration as mentioned before.

sharonyang replied:
Done.

lakshbhasin commented on network_simulator/plot_tool.py:
```diff
@@ -0,0 +1,300 @@
+
+            # Packet RTT times.
+            plt.subplot(438)
+            self.tool.graph_1d_list(flow_stats.packet_rtts)
```
Packet RTTs is a plot of RTT (in sec) vs time (in sec) when that RTT was received. It is not a 1D plot. You might have to change the "flow_packet_received" function and the description for "packet_rtts" so that packet_rtts is actually a list of (timestamp, RTT) tuples, where timestamp is the curr_time when the RTT was received.

sharonyang replied:
Done.

lakshbhasin commented on network_simulator/plot_tool.py:
```diff
@@ -0,0 +1,300 @@
+        plt.legend(sent_legend)
+        # Set up labels.
+        plt.xlabel("Time (sec)")
+        plt.ylabel("Packet Sent Rate")
```
Call this the "Flow send rate (Mbps)". Again, the units for this is Mbps. You can convert a rate (in bps) into Mbps by dividing by MEGABIT.

sharonyang replied:
Done.

lakshbhasin commented on network_simulator/plot_tool.py:
```diff
@@ -0,0 +1,300 @@
+        plt.xlabel("Time (sec)")
+        plt.ylabel("Packet Sent Rate")
+        # Set up grid on graph.
+        plt.grid(True)
```
Again, no grids anywhere.

sharonyang replied:
Done.

lakshbhasin commented on network_simulator/plot_tool.py:
```diff
@@ -0,0 +1,300 @@
+        plt.legend(receive_legend)
+        # Set up labels.
+        plt.xlabel("Time (sec)")
+        plt.ylabel("Packet Received Rate")
```
Call this the "Flow receive rate (Mbps)", i.e. with units of Mbps. Divide rate data by MEGABIT in the for loop above.

sharonyang replied:
Done.

lakshbhasin commented on network_simulator/plot_tool.py:
```diff
@@ -0,0 +1,300 @@
+        Graphs the host stats of each available host.
+        """
+        # Set up legend storage for host names.
+        sent_legend = []
```
Comment before each legend describing what stats are being looked at.

sharonyang replied:
Done.

lakshbhasin commented on network_simulator/plot_tool.py:
```diff
@@ -0,0 +1,300 @@
+
+        # Iterate through each host to display each stats.
+        for host, host_stats in self.stats.host_stats.iteritems():
+            # Name of host would be "host_address".
```
nit: Name of host is its "address" field.

sharonyang replied:
Done.

lakshbhasin commented on network_simulator/plot_tool.py:
```diff
@@ -0,0 +1,300 @@
+            # Name of host would be "host_address".
+            # Packet sent times.
+            plt.subplot(433)
+            sent_rate_lst = self.tool.gen_rate_tuple_list(
```
Divide all rates' y-axes by MEGABIT as mentioned before (this should be done for all rates in other functions too). Something like sent_rate_lst = [(a[0], float(a[1]) / MEGABIT) for a in sent_rate_lst] should work I think.

sharonyang replied:
Done.

lakshbhasin commented on network_simulator/plot_tool.py:
```diff
@@ -0,0 +1,300 @@
+        plt.legend(sent_legend)
+        # Set up labels.
+        plt.xlabel("Time (sec)")
+        plt.ylabel("Packet Sent Rate")
```
Call this the "Host send rate (Mbps)".

sharonyang replied:
Done.

lakshbhasin commented on network_simulator/plot_tool.py:
```diff
@@ -0,0 +1,300 @@
+        plt.legend(receive_legend)
+        # Set up labels.
+        plt.xlabel("Time (sec)")
+        plt.ylabel("Packet Received Rate")
```
Call this the "Host receive rate (Mbps)".

sharonyang replied:
Done.

lakshbhasin commented on network_simulator/plot_tool.py:
```diff
@@ -0,0 +1,300 @@
+
+    def graph_network(self):
+        """
+        Graphs the entire network (this includes all links, flows, and hosts.
```
nit: close parenthesis.

sharonyang replied:
Done.

lakshbhasin commented on network_simulator/plot_tool.py:
```diff
@@ -0,0 +1,300 @@
+
+        # Make spacing between plots.
+        plt.tight_layout()
+        #plt.axis('equal')
```
Could remove this line, or have an inline comment describing what it used to do (inline comments should occur two spaces after the code ends, i.e. "a = foo  # bar").

sharonyang replied:
Done.

lakshbhasin commented on network_simulator/statistics.py:
```diff
@@ -181,7 +180,18 @@ def flow_packet_received(self, flow, ack_packet, curr_time):
         # it to calculate RTT.
         sent_time = ack_packet.data_packet_start_time_sec
-        stats.packet_rtts.append(curr_time - sent_time)
+        stats.packet_rtts.append(
```
In the first line of this function, assert ack_packet instanceof AckPacket.

sharonyang replied:
Done.

lakshbhasin replied:
packet_rtts is a list of (timestamp, RTT) tuples, so what you want to record is (curr_time, curr_time - sent_time)

sharonyang replied:
Done.

lakshbhasin commented on network_simulator/statistics.py:
```diff
@@ -166,7 +164,8 @@ def flow_packet_sent(self, flow, data_packet):
         """
         stats = self.get_flow_stats(flow)
-        stats.packet_sent_times.append(data_packet.start_time_sec)
+        stats.packet_sent_times.append(
```
In the first line of this function, assert data_packet instanceof DataPacket.

sharonyang replied:
Done.

lakshbhasin commented on network_simulator/statistics.py:
```diff
@@ -181,7 +180,18 @@ def flow_packet_received(self, flow, ack_packet, curr_time):
         sent_time = ack_packet.data_packet_start_time_sec
-        stats.packet_rtts.append(curr_time - sent_time)
+        stats.packet_rtts.append(
+            (curr_time - sent_time, ACK_PACKET_SIZE_BITS))
```
You can use ack_packet.size_bits here too.

sharonyang replied:
Done.

lakshbhasin commented on network_simulator/statistics.py:
```diff
@@ -34,18 +34,20 @@ class FlowStats(object):
+    tuples. packet_size is in bits. Time is in seconds.
     :ivar list packet_rtts: a list of RTTs (in seconds).
+    :ivar list window_size_times: a list of (timestamp, window_size) with
+    window_size in number of packets.
```
nit: mention time is in seconds.

sharonyang replied:
Done.

lakshbhasin commented on network_simulator/statistics.py:
```diff
@@ -34,18 +34,20 @@ class FlowStats(object):
+                 packet_rtts=[], window_size_times=[]):
         packet_sent_times = packet_sent_times
         packet_rec_times = packet_rec_times
         packet_rtts = packet_rtts
```
As I'd mentioned earlier, packet_rtts should actually be a list of (timestamp, rtt) tuples, so we can track how the RTT changes over time for each flow.

sharonyang replied:
Done.

lakshbhasin commented on network_simulator/plot_tool.py:
```diff
@@ -0,0 +1,300 @@
+        trans_legend = []
+
+        # Iterate through each link to display each stats.
+        for link, link_stats in self.stats.link_stats.iteritems():
```
Is "link" actually link_name?

sharonyang replied:
Done.

lakshbhasin commented on network_simulator/plot_tool.py:
```diff
@@ -0,0 +1,300 @@
+        window_legend = []
+
+        # Iterate through each flow to display each stats.
+        for flow, flow_stats in self.stats.flow_stats.iteritems():
```
Isn't "flow" actually "flow_id"?

sharonyang replied:
Done.

lakshbhasin commented on network_simulator/plot_tool.py:
```diff
@@ -0,0 +1,300 @@
+        receive_legend = []
+
+        # Iterate through each host to display each stats.
+        for host, host_stats in self.stats.host_stats.iteritems():
```
Isn't "host" actually "host_addr"?

sharonyang replied:
Done.

lakshbhasin commented on network_simulator/statistics.py:
```diff
@@ -181,7 +180,18 @@ def flow_packet_received(self, flow, ack_packet, curr_time):
+        stats.packet_rtts.append(
+            (curr_time - sent_time, ACK_PACKET_SIZE_BITS))
+
+    def flow_window_size(self, flow, curr_time):
```
Add a TODO for me to incorporate this into flow.py code.

sharonyang replied:
Done.

sharonyang committed:
Modified statistics.py, plot_tools.py per Laksh's comments #1 for
PR #16.

sharonyang committed:
Output averages of each Device to stdout.

sharonyang committed:
Fixed bugs after attempting to run the plotting tool with random
values.

This includes some misunderstanding in static class/methods set up
and comment changes.

lakshbhasin commented on network_simulator/plot_tool.py:
```diff
@@ -0,0 +1,390 @@
+            timestamp = lst[idx]
+            if timestamp > curr_window_end_time:
+                output.append((curr_window_end_time, curr_window_count))
+                curr_count = 0
```
I think you meant to set curr_window_count = 0. Be careful about renaming in Python (use PyCharm's refactor where possible).

sharonyang replied:
Done.

lakshbhasin commented on network_simulator/plot_tool.py:
```diff
@@ -0,0 +1,390 @@
+                output.append((curr_window_end_time, curr_window_count))
+                curr_count = 0
+                curr_window_end_time += GRAPH_WINDOW_SIZE
+            if timestamp <= curr_window_end_time:
```
Yeah, you're right, having this as a separate if will work.

sharonyang replied:
Done.

lakshbhasin commented on network_simulator/statistics.py:
```diff
@@ -16,8 +16,8 @@ class LinkStats(object):
     tuples.
-    :ivar list packet_loss_times: a list of timestamps, each corresponding
-    to a packet loss.
+    :ivar list packet_loss_times: a list of (timestamps, packet_size), each
```
Isn't this just a list of timestamps? I think we just count number of packet losses, not the size of them.

sharonyang replied:
Done.

lakshbhasin commented on network_simulator/plot_tool.py:
```diff
@@ -0,0 +1,390 @@
+
+class Grapher(object):
+    """
+    Graphs statistics items from Statistics class.
```
Maybe mention this also outputs averages (and change the name from "Grapher" to "Analyzer" or something to reflect this?)

sharonyang replied:
Done.

lakshbhasin commented on network_simulator/plot_tool.py:
```diff
@@ -0,0 +1,390 @@
+            else:
+                latest_timestamp = tuple_list[len(tuple_list) - 1][0]
+                avg = float(curr_sum) / latest_timestamp
+        print id + " Average " + value_type + ": " + str(avg) + " " + units
```
You should import logging and set "logger = logging.getLogger(__name__)" at the top of the file (e.g. see network_topology.py). Then, use logger.info() wherever possible, instead of "print". logger.info can work with strings as you've passed in, or with format strings:

logger.info("%s Average %s: %f %s", id, value_type, avg, units)

or something like that.

sharonyang replied:
Done.

lakshbhasin commented on network_simulator/plot_tool.py:
```diff
@@ -0,0 +1,390 @@
+        if count != 0:
+            latest_timestamp = tuple_list[len(tuple_list) - 1][0]
+            avg = float(count) / latest_timestamp
+        print id + " Average " + value_type + ": " + str(avg) + " " + units
```
Same thing about using logger.info() as before.

sharonyang replied:
Done.

lakshbhasin commented on network_simulator/plot_tool.py:
```diff
@@ -0,0 +1,390 @@
+            plt.plot(*zip(*tuple_list), marker='o', linestyle='--')
+
+    @staticmethod
+    def output_sum_avg_tuple_list(id, value_type, count, units, tuple_list=[]):
```
I don't think you need to set tuple_list=[] as a default value?

sharonyang replied:
Done

lakshbhasin commented on network_simulator/plot_tool.py:
```diff
@@ -0,0 +1,390 @@
+            if count:
+                avg = float(curr_sum) / len(tuple_list)
+            else:
+                latest_timestamp = tuple_list[len(tuple_list) - 1][0]
```
nit: you can replace len(tuple_list) - 1 with just -1 because of how Python indexing works.

sharonyang replied:
Done

lakshbhasin commented on network_simulator/plot_tool.py:
```diff
@@ -0,0 +1,390 @@
+            occpy_legend.append(link_name)
+            # Output average buffer occupancy to stdout.
+            PlotTool().output_sum_avg_tuple_list(
+                "Link " + link_name,
```
You should name variables in this function call, i.e. id="Link " + link_name, value_type="Buffer Occupancy", and so on. Same with the other calls involving many function parameters.

sharonyang replied:
Done.

lakshbhasin commented on network_simulator/plot_tool.py:
```diff
@@ -0,0 +1,390 @@
+        Outputs the average value of a tuple list. This is (count / time).
+
+        Assume that a tuple list is in the format of (timestamp, data), we
+        iterature through the data field of the tuples and outputs the
```
nit: iterate :)

sharonyang replied:
Done.

lakshbhasin commented on network_simulator/plot_tool.py:
```diff
@@ -0,0 +1,390 @@
+        count (True), or a timestamp value (False).
+        :param string units: units of the average. e.g. "(sec)"
+        :param list tuple_list: the tuple list to be computed.
+        :return: no return, just print out the statement to stdout.
```
change this to "just log the average" once you've moved to the logger package.

sharonyang replied:
Done.

lakshbhasin commented on network_simulator/plot_tool.py:
```diff
@@ -0,0 +1,390 @@
+        :param string value_type: the data type of the average.
+        :param string units: units of the average. e.g. "(sec)"
+        :param list tuple_list: the tuple list to be computed.
+        :return: no return, just print out the statement to stdout.
```
same thing about mentioning use of logger.

sharonyang replied:
Done.

lakshbhasin commented on network_simulator/plot_tool.py:
```diff
@@ -0,0 +1,390 @@
+        count = len(tuple_list)
+        avg = 0
+        if count != 0:
+            latest_timestamp = tuple_list[len(tuple_list) - 1][0]
```
same thing about using -1 to access the last element of a list.

sharonyang replied:
Done.

lakshbhasin commented on network_simulator/plot_tool.py:
```diff
@@ -0,0 +1,390 @@
+    @staticmethod
+    def output_count_avg_tuple_list(id, value_type, units, tuple_list=[]):
+        """
+        Outputs the average value of a tuple list. This is (count / time).
```
just to make it clear, add that count = len(tuple_list) or something.

sharonyang replied:
Done.

lakshbhasin commented on network_simulator/plot_tool.py:
```diff
@@ -0,0 +1,390 @@
+            # Name of link would be "link_name".
+            # Buffer occupancy w.r.t time.
+            plt.subplot(431)
+            PlotTool().graph_tuple_list(link_stats.buffer_occupancy)
```
You should use "PlotTool" instead of "PlotTool()", since the latter constructs an object (which you don't need). This should be changed in all the other cases below as well.

sharonyang replied:
Done.

lakshbhasin commented on network_simulator/plot_tool.py:
```diff
@@ -0,0 +1,390 @@
+            PlotTool().graph_tuple_list(flow_rate_list)
+            trans_legend.append(link_name)
+            # Output average packet transmitted per sec to stdout.
+            PlotTool().output_count_avg_tuple_list(
```
We actually want to output the average flow rate on the link in Mbps. To do this, you need to just sum up all of the data sent (in megabits) and divide by the biggest timestamp in order to get a rate in bps. Then divide by MEGABIT.

sharonyang replied:
Done.

lakshbhasin commented on network_simulator/plot_tool.py:
```diff
@@ -0,0 +1,390 @@
+            plt.subplot(431)
+            PlotTool().graph_tuple_list(link_stats.buffer_occupancy)
+            occpy_legend.append(link_name)
+            # Output average buffer occupancy to stdout.
```
change "stdout" to "log" (or something) in all of the comments that are like this.

sharonyang replied:
Done.

lakshbhasin commented on network_simulator/plot_tool.py:
```diff
@@ -0,0 +1,390 @@
+            PlotTool().graph_tuple_list(sent_rate_lst)
+            sent_legend.append(flow_id)
+            # Output average packet sent per sec to stdout.
+            PlotTool().output_count_avg_tuple_list(
```
Again, we want to compute the average send rate in Mbps, so you need to sum packet sizes in bits, divide by the largest timestamp, and then divide by MEGABIT.

sharonyang replied:
Done.

lakshbhasin commented on network_simulator/plot_tool.py:
```diff
@@ -0,0 +1,390 @@
+            PlotTool().graph_tuple_list(rec_rate_lst)
+            receive_legend.append(flow_id)
+            # Output average packet received per sec to stdout.
+            PlotTool().output_count_avg_tuple_list(
```
Same comment as mentioned above.

sharonyang replied:
Done.

lakshbhasin commented on network_simulator/plot_tool.py:
```diff
@@ -0,0 +1,390 @@
+            PlotTool().graph_tuple_list(sent_rate_lst)
+            sent_legend.append(host_addr)
+            # Output average packet received per sec to stdout.
+            PlotTool().output_count_avg_tuple_list(
```
Same as above; need to output the average send rate (for this host) in Mbps. Also make sure to update the comments.

sharonyang replied:
Done.

lakshbhasin commented on network_simulator/plot_tool.py:
```diff
@@ -0,0 +1,390 @@
+            PlotTool().graph_tuple_list(rec_rate_lst)
+            receive_legend.append(host_addr)
+            # Output average packet received per sec to stdout.
+            PlotTool().output_count_avg_tuple_list(
```
Same as above (need to output the average receipt rate for this host in Mbps).

sharonyang replied:
Dne.

lakshbhasin commented on network_simulator/statistics.py:
```diff
@@ -34,18 +34,21 @@ class FlowStats(object):
     def __init__(self, packet_sent_times=[], packet_rec_times=[],
-                 packet_rtts=[]):
+                 packet_rtts=[], window_size_times=[]):
         packet_sent_times = packet_sent_times
```
Instance variables need to be set via self, e.g. self.packet_sent_times = .... PyCharm should mark this as an issue in some way. This should be fixed in HostStats too.

sharonyang replied:
Done

lakshbhasin commented on network_simulator/statistics.py:
```diff
@@ -69,8 +72,7 @@ class Statistics(object):
     Intended to be owned by the main loop to record all statistics within
```
nit: gap between classes (three lines above this one) should be only 2 lines long.

sharonyang replied:
Done

sharonyang committed:
Updated PR #16 per Laksh's comments 2.

lakshbhasin commented on network_simulator/plot_tool.py:
```diff
@@ -0,0 +1,424 @@
+            # Output average packet transmitted per sec to log.
+            PlotTool.output_rate_avg_tuple_list(
+                id="Link " + link_name,
+                value_type="Packet Transmitted", units="(Mbps)",
```
value_type should be "Transmission Rate"

sharonyang replied:
Done

lakshbhasin commented on network_simulator/plot_tool.py:
```diff
@@ -0,0 +1,424 @@
+            # Output average packet sent per sec to log.
+            PlotTool.output_rate_avg_tuple_list(
+                id="Flow " + flow_id,
+                value_type="Packet Sent", units="(Mbps)",
```
value_type should be "Send Rate"

sharonyang replied:
Done.

lakshbhasin commented on network_simulator/plot_tool.py:
```diff
@@ -0,0 +1,424 @@
+            # Output average packet received per sec to log.
+            PlotTool.output_rate_avg_tuple_list(
+                id="Flow " + flow_id,
+                value_type="Packet Received", units="(Mbps)",
```
value_type should be "Receive Rate"

sharonyang replied:
Done.

lakshbhasin commented on network_simulator/plot_tool.py:
```diff
@@ -0,0 +1,424 @@
+            # Output average packet received per sec to log.
+            PlotTool.output_rate_avg_tuple_list(
+                id="Host " + host_addr,
+                value_type="Packet Sent", units="(Mbps)",
```
value_type should be "Send Rate"

sharonyang replied:
Done.

lakshbhasin commented on network_simulator/plot_tool.py:
```diff
@@ -0,0 +1,424 @@
+            # Output average packet received per sec to log.
+            PlotTool.output_rate_avg_tuple_list(
+                id="Host " + host_addr,
+                value_type="Packet Received", units="(Mbps)",
```
value_type should be "Receive Rate"

sharonyang replied:
Done.

lakshbhasin commented:
Looks great, just some minor changes and it should be good to merge. We will need to test that this is working @yubo56 @dluo11.

sharonyang committed:
Changed average naming per Laksh's comment #3 for PR #16.
